### PR TITLE
[hw] Support `spark_config` in pipeline `metadata.yaml`

### PR DESCRIPTION
**Pull request by csharplus**
# Summary

Currently custom `spark_config` is only supported at the project (repo) level, which has been enhanced to be configurable at the pipeline level.

# Tests

Manually tested in the local development environment, with the following procedures. For the default Mage development environment, PySpark is not installed, which could be added by using the following procedure:

* Click the Terminal icon on the right side of the Mage UI, and install the required packages and libraries for pyspark:
```
root@488dc9529cf3:/home/src#  apt-get update && \
apt-get install -y --no-install-recommends \
        openjdk-11-jre

root@488dc9529cf3:/home/src#  pip install pyspark
```

1. Add the following configuration to the `metadata.yaml` under a newly created pipeline:
```
spark_config:
  app_name: my spark app 10
  executor_env: {}
  others: {}
  spark_home: null
  spark_jars: []
  spark_master: local
```
Screenshot 2023-06-06 at 1 46 52 AM

The `spark_config` in `metadata.yaml` under the project (i.e., Repo) folder has a different setting for `app_name`, i.e.,  `app_name: my spark app` instead of  `app_name: my spark app 10`,  as shown below:
Screenshot 2023-06-06 at 1 49 22 AM

2. Add a data loader block with the following code:
```
@data_loader
def load_data(*args, **kwargs):
    """
    Template code for loading data from any source.

    Returns:
        Anything (e.g. data frame, dictionary, array, int, str, etc.)
    """
    # Specify your data loading logic here

    spark = kwargs.get('spark')
    if spark:
        spark_conf = spark.sparkContext.getConf()
        print('#' * 40)
        print(spark_conf.toDebugString())
        print('#' * 40)
        print(spark.sql('select 1'))
        return {}
    else:
        print("No active spark found")

    return 1
```
Screenshot 2023-06-06 at 1 47 56 AM

3. Run the data loader by clicking the `play` button. From the results of the test run, it was confirmed that the pipeline `metadata.yaml` was used for Spark Config, as shown in `spark.app.name=my spark app 10`:
```
1/1 tests passed.
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/06/06 08:37:47 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/06/06 08:37:48 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
########################################
spark.app.id=local-1686040669084
spark.app.name=my spark app 10
spark.app.startTime=1686040667735
spark.app.submitTime=1686040667508
spark.driver.extraJavaOptions=-Djava.net.preferIPv6Addresses=false -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -Djdk.reflect.useDirectMethodHandle=false
spark.driver.host=************
spark.driver.port=37825
spark.executor.extraJavaOptions=-Djava.net.preferIPv6Addresses=false -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -Djdk.reflect.useDirectMethodHandle=false
spark.executor.id=driver
spark.master=local
spark.rdd.compress=True
spark.serializer.objectStreamReset=100
spark.submit.deployMode=client
spark.submit.pyFiles=
spark.ui.showConsoleProgress=true
########################################
DataFrame[1: int]
{}
```
Screenshot 2023-06-06 at 1 55 23 AM

4. If changing the pipeline `metadata.yaml` setting, e.g., from `app_name: my spark app 10` to `app_name: my spark app 20`. Executing the same data loader again would lead to a different Spark Session Conf, as shown below:
Screenshot 2023-06-06 at 1 54 31 AM

5. Create three different pipelines in one project with a `Data Loader` block in each pipeline, with the following code:
```
    spark = kwargs.get('spark')
    if spark:
        spark_conf = spark.sparkContext.getConf()
        print('#' * 40)
        print(spark_conf.toDebugString())
        print('#' * 40)
        print(spark.sql('select 1'))
        return {}
    else:
        print("No active spark found")
```

6. Set `spark_config` in pipeline's `metadata.yaml` in three different ways:
* With empty `spark_config` setting in pipeline's `metadata.yaml`
* Choose  a different `app_name` in `spark_config` setting in pipeline's `metadata.yaml` from that in project's ``metadata.yaml`
* Choose  another different `app_name` in `spark_config` setting in pipeline's `metadata.yaml` from both that in project's ``metadata.yaml' and that of the previous pipeline 

7. Execute these pipeline once manually in the `Triggers` tab, and check the `Logs` table,  which would verify different `Spark Conf` details were printed out as expected for each pipeline.

cc:

@dy46 @wangxiaoyou1993

### CODE REVIEW

File: mage_ai/services/spark/config.py [hunk 1]
```diff
@@ -4,4 +4,4 @@
 
 @dataclass
 class SparkConfig(BaseConfig):
-    repo_path: str = None
+    spark_config: dict = None
```
Comment (wangxiaoyou1993) [inline mage_ai/services/spark/config.py:7]: why don't we put the config fields in this dataclass?
for example
```
@dataclass
class SparkConfig(BaseConfig):
    app_name: str = None
    spark_master: str = None
    spark_home: str = None
    executor_env: Dict = field(default_factory=dict)
    spark_jars: List = field(default_factory=list)
    others: Dict = field(default_factory=dict)
```

Comment (csharplus) [reply mage_ai/services/spark/config.py]: Good idea: made the suggested change.

File: mage_ai/data_preparation/models/pipeline.py [hunk 2]
```diff
@@ -128,6 +127,10 @@ def version(self):
     def version_name(self):
         return f'v{self.version}'
 
+    @property
+    def pipeline_spark_config(self):
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/pipeline.py:131]: do you need this property method? is it the same with self.spark_config?

File: mage_ai/data_preparation/repo_manager.py [hunk 3]
```diff
@@ -114,6 +117,10 @@ def from_dict(self, config_dict: Dict) -> 'RepoConfig':
     def metadata_path(self) -> str:
         return os.path.join(self.repo_path, 'metadata.yaml')
 
+    @property
+    def repo_spark_config(self) -> dict:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/repo_manager.py:121]: Is this method same with self.spark_config?

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_ai/data_preparation/models/pipeline.py (hunk 2 above)
Comment (csharplus) [reply mage_ai/data_preparation/models/pipeline.py]: Good catch: removed it.

File: mage_ai/data_preparation/repo_manager.py (hunk 3 above)
Comment (csharplus) [reply mage_ai/data_preparation/repo_manager.py]: Right, removed this property to avoid duplication.

**Pull request merged**